prov/mrail/src/mrail_init.c

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (c) 2018 Intel Corporation, Inc.  All rights reserved.
+ *
+ * This software is available to you under a choice of one of two
+ * licenses.  You may choose to be licensed under the terms of the GNU
+ * General Public License (GPL) Version 2, available from the file
+ * COPYING in the main directory of this source tree, or the
+ * BSD license below:
+ *
+ *     Redistribution and use in source and binary forms, with or
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include "mrail.h"
+
+int mrail_get_core_info(uint32_t version, const char *node, const char *service,
+			uint64_t flags, const struct fi_info *hints,
+			struct fi_info **core_info)
+{
+	struct fi_info *core_hints;
+	const char *core_name;
+	size_t len;
+	int ret;
+
+	if (!(core_hints = fi_dupinfo(hints)))
+		return -FI_ENOMEM;
+
+	free(core_hints->fabric_attr->prov_name);
+
+	core_name = ofi_core_name(hints->fabric_attr->prov_name, &len);
+	if (core_name) {
+		core_hints->fabric_attr->prov_name = strndup(core_name, len);
+		if (!core_hints->fabric_attr->prov_name) {
+			FI_WARN(&mrail_prov, FI_LOG_FABRIC,
+				"Unable to alloc prov name\n");
+			ret = -FI_ENOMEM;
+			goto out;
+		}
+	}
+	flags |= OFI_CORE_PROV_ONLY;
```

Comment:
We'll eventually want to support multi-rail over the utility providers.  So, we'll need some other sort of way to handle this.  Tossing out an idea, maybe we need a layer-number that we're safe to layer over.  Core=layer 0, utility=layer 1, multi-rail is layer 2